src/lib/components/Menu/Menu.js

```diff
@@ -77,129 +78,350 @@ const mapStateToProps = (state, ownProps) => {
+  };
+
+  componentDidUpdate(prevProps) {
+    if (this.props.showMap && this.props.showMap !== prevProps.showMap && this.props.menuScroll) {
+      this.menuWrapper.current.scrollTop = this.props.menuScroll.scrollTop;
     }
   }
 
-  onToggleMenu = (e) => {
+  onToggleMenu = e => {
     e.preventDefault();
     const { dispatch } = this.props;
     dispatch(Actions.toggleMenu(this.props.mapId));
-  }
+  };
 
   onCategoryClick = (e, category) => {
     e.preventDefault();
     const { openCategories } = this.props;
     const index = openCategories.indexOf(category);
-    this.props.dispatch(Actions.toggleCategories(this.props.mapId, category, index))
-  }
+    this.props.dispatch(Actions.toggleCategories(this.props.mapId, category, index));
+  };
 
-  onRegionClick = (e) => {
+  onRegionClick = e => {
     const region = e.target.value;
     this.props.dispatch(Actions.changeRegion(region));
+  };
+
+  /**
+   * Modify a layer by giving its group and nested groups a count
+   * @param {Object} layer - A layer with groups
+   * @param {number} groupCounter - A group counter that will give each group a number
+   * @returns {Object} Modified layer and current group counter increment
+   */
+  insertGroupCount(layer, groupCounter) {
+    Object.keys(layer).forEach(key => {
+      layer[key].count = groupCounter;
+      groupCounter += 1;
+
+      layer[key].layers.forEach(keyLayer => {
+        if (!keyLayer.id) {
+          keyLayer = this.insertGroupCount(keyLayer, groupCounter);
+        }
+      });
+    });
+
+    return {
+      layer,
+      groupCounter,
+    };
+  }
+
+  /**
+   * Loop over the categories and modify data
+   * @param {array} categories Menu categories and their layers, basically all menu items
+   * @returns {array} Modified categories
+   */
+  parseCategories(categories) {
+    let groupCounter = 0;
+    categories.forEach(category => {
+      category.layers.forEach(layer => {
+        if (!layer.id) {
+          // This is a group. Give it and its nested groups a count
+          // This will make sure when we toggle a group, another group with the
+          // exact same name is not toggled. The array of open groups is stored as an
+          // array in the store. Another implementaion would to give each group an Id e.g a random no but
+          // this means different menu instances will have groups that have different Ids which we do not
+          // want (We would like menu instances under the same map ID to behave the same. If say I open a group in one
+          // menu, it should appear open the other ). A contigous int counter is
+          // therefore used to ensure the counter for a group in different menu
+          // instances is the same, and also, groups in the same menu instances do not share a count.
+          const obj = this.insertGroupCount(layer, groupCounter);
+          layer = obj.layer;
+          groupCounter = obj.groupCounter;
```

Comment:
Why are we using `groupCounter` here instead of the second param (the iterator) of the inner `forEach` loop? Is it not enough to memonize the categories?

Also lets not mutate the categories being passed in through the params, and instead use new category objects to build a new `nextCategories` object which is returned. In this way, everything that happens in this codeblock should be easily swept up by the garbage collector (instead of managing/passing along references).

Reply:
> Why are we using `groupCounter` here instead of the second param (the iterator) of the inner `forEach` loop? Is it not enough to memonize the categories?

@cKellyDesign  We are using the group counter because we would like to give each group a contiguous number that is not shared with another group. That number should also match the number of the same group in another menu instance e.g superset menu (superset at layer level for irfsomalia). If I have a group with group number 1 on the map, then navigate to the data view, the same group should have a group counter of 1.

Why did we decide to use redux state to keep track of open groups and not the previous implementation of local state? In irfsomalia, we wanted to have the same menu group open if a user switches to data view.

Why did we introduce group counters? The initial implementation of using the redux state to keep track of open groups was absolutely similar to the one used to keep track of open categories. However, we experienced a problem. A group in a category that shared a name with another group in another category e.g **Output Indicators** would cause other output indicators to open if clicked open or force them to close if it was closed. We modified the state again to keep track of an open group by storing its **group counter** and **name**.

The solutions to the problems above are welcome to further enhancement or changed entirely completely. A better solution I think would be to have one instance of Menu (as a child of AppView) and remove it as a child of Map and display/hide according to needs.

For the performance, I have overridden `ComponentDidUpdate`

Reply:
Thanks @kelvin-muchiri for the detailed response! 🙌  
> Why did we decide to use redux state to keep track of...

I definitely think it makes sense to have a redux component to keep track of things for the component in a modular and decoupled fashion.

If this:
> A group in a category that shared a name with another group in another category e.g Output Indicators would cause other output indicators to open if clicked open or force them to close if it was closed.

Is happening:
>because categories and groups don't have IDs

Could we generate IDs for the categories such that we could get Categories by IDs (similar to the stores in Reveal)?

> Could we generate IDs...

Would this be easier than parsing categories? Could it break other things?